projects/openapi-utilities/src/diff/diff.ts

```diff
@@ -0,0 +1,205 @@
+
+    // Start by matching up values to compare
+    // for arrays, we need to match up
+    if (Array.isArray(before.value) && Array.isArray(after.value)) {
+      const allValues = [...before.value, ...after.value];
+      const arrayIdFn: (v: any, i: number) => string = allValues.every((v) =>
+        isParameterObject(v)
+      )
+        ? getParameterIdentity
+        : allValues.every((v) => typeof v !== 'object')
+        ? (v: any) => String(v)
+        : (_, i: number) => String(i);
+
+      const beforeValuesById: Map<string, [JSONValue, number]> = new Map(
+        before.value.map((v, i) => [arrayIdFn(v, i), [v, i]])
+      );
+      const afterValuesById: Map<string, [JSONValue, number]> = new Map(
+        after.value.map((v, i) => [arrayIdFn(v, i), [v, i]])
+      );
+
+      const keys = new Set([
+        ...beforeValuesById.keys(),
+        ...afterValuesById.keys(),
+      ]);
+      for (const key of keys) {
+        const [beforeValue, beforeIdx] = beforeValuesById.get(key) ?? [];
+        const [afterValue, afterIdx] = afterValuesById.get(key) ?? [];
+        const beforePath = jsonPointerHelpers.append(
+          before.path,
+          String(beforeIdx)
+        );
+        const afterPath = jsonPointerHelpers.append(
+          after.path,
+          String(afterIdx)
+        );
+
+        comparisons.push({
+          beforeValue,
+          beforePath,
+          afterValue,
+          afterPath,
+        });
+      }
+    } else if (!Array.isArray(before.value) && !Array.isArray(after.value)) {
+      const keys = new Set([
+        ...Object.keys(before.value),
+        ...Object.keys(after.value),
+      ]);
+
+      for (const key of keys) {
+        const beforeValue = before.value[key];
+        const beforePath = jsonPointerHelpers.append(before.path, key);
+        const afterValue = after.value[key];
+        const afterPath = jsonPointerHelpers.append(after.path, key);
+        comparisons.push({
+          beforeValue,
+          beforePath,
+          afterValue,
+          afterPath,
+        });
+      }
+    } else {
+      throw new Error(
+        'Unexpectedly found mismatch between array and object in diff traversal'
+      );
+    }
+
+    for (const {
+      beforeValue,
+      beforePath,
+      afterValue,
+      afterPath,
+    } of comparisons) {
+      if (beforeValue && afterValue === undefined) {
+        // We don't need to look at the last key to see path differences
+        const beforeParts = jsonPointerHelpers.decode(beforePath).slice(0, -1)
+        const afterParts = jsonPointerHelpers.decode(afterPath).slice(0, -1)
+        const pathReconciliation: [number, string][] = [];
+        for( let i = 0; i < beforeParts.length; i++) {
+          const before = beforeParts[i];
+          const after = afterParts[i];
+          if (before !== after) {
+            pathReconciliation.push([i, after]);
+          }
+        }
```

Comment:
See test case https://github.com/opticdev/optic/pull/1439/files#diff-e592c7df8403844274bc28cdb28360de2572020a2fe6c638df9bd70490d7280aR22 for example and https://www.notion.so/useoptic/Optimize-Changelog-Results-File-8fcd111e08904cf882af6b7a0103f977#d41d1ce10746419da30d6678824b8add for details